---
title: scATAC-seq Report
date: "`r Sys.Date()`"
output:
  flexdashboard::flex_dashboard:
    vertical_layout: scroll
    orientation: rows
    theme:
      version: 4
      bootswatch: sandstone
params:
  nCPUs: 6
  nMem: 10000000000.0
  sampleName: NULL
  stats_tsv: NULL
  raw_cells: NULL
  raw_meta: NULL
  obj_filtered: NULL
  macs_peaks: NULL
  saturation_json: NULL
---

```{r include=FALSE}
library(tidyverse)
library(scales)
library(kableExtra)
library(flexdashboard)
library(Signac)
library(Seurat)
##library(GenomeInfoDb)
##library(EnsDb.Mmusculus.v79)
library(rtracklayer)
library(patchwork)
library(future)
library(plotly)
library(jsonlite)
library(reticulate)

set.seed(1234)
plan("multisession", workers = as.numeric(params$nCPUs))
options(future.globals.maxSize = as.numeric(params$nMem))
RhpcBLASctl::blas_set_num_threads(1) # https://github.com/satijalab/seurat/issues/3991


## Rmd options
knitr::opts_chunk$set(
  comment = '', fig.retina = 4,
  warning = FALSE, message = FALSE
)

config_plotly_fig <- function(fig){
    config(
        fig,
        displaylogo = FALSE, 
        modeBarButtonsToRemove = c('zoom', 'pan', 'select', 'zoomIn', 'zoomOut', 'autoScale',
                                   'hoverClosestCartesian', 'hoverCompareCartesian'),
        toImageButtonOptions = list(height= NULL, width= NULL, scale= 2)
    )
}
```

```{css echo = FALSE}
.chart-title {
  font-size: 1.2rem;
  font-weight: 700;
}
```

```{js echo=FALSE}
document.querySelector(".navbar").classList.remove("navbar-inverse");
document.querySelector(".navbar").classList.add("navbar-default");
document.querySelector(".navbar").classList.add("pt-1");
document.querySelector(".navbar").classList.add("pb-1");
```

```{r preprocess, include=FALSE}
obj_filtered <- readRDS(params$obj_filtered)
obj_filtered$orig.ident <- params$sampleName

tssPlot <- TSSPlot(obj_filtered, group.by = 'high.tss') + NoLegend()

## Chr1 fragments length distribution
fragLenPlot <- FragmentHistogram(object = obj_filtered, group.by = 'nucleosome_group', region=paste0(seqlevels(Annotation(obj_filtered))[1],'-1-200000000'))

qcPlot <- VlnPlot(
  object = obj_filtered,
  features = c('FRiP', 'nCount_peaks',
               'TSS.enrichment', 'blacklist_fraction', 'nucleosome_signal'),
  pt.size = 0.1,
  ncol = 5,
  group.by = "orig.ident"
)


## Filter data with otsu threshold of nCount_peaks, and with other naive criterias
cells <- colnames(obj_filtered)

raw_cells <- read_tsv(params$raw_cells, col_names = FALSE) %>% pull()

d <- read_tsv(params$raw_meta) ## counts = fragsOverlapPeaks

d <- d %>% dplyr::arrange(desc(fragsOverlapPeaks)) %>%
    dplyr::mutate(idx=1:nrow(d))

## Do not perform clustering when there are too few cells
if(length(cells)>=150){
    clusterPlot <- DimPlot(object = obj_filtered, label = TRUE) + NoLegend()
}
```

```{r read_stats, include = FALSE}
stats <- read_tsv(params$stats_tsv, col_names=FALSE)
colnames(stats) <- c("Term", "value")
## Add more stats about cells
cellFrag_median <- obj_filtered[["fragments"]] %>% pull(fragments) %>% median() %>% as.integer()
cellFrag_mean <- obj_filtered[["fragments"]] %>% pull(fragments) %>% mean() %>% as.integer()
cellFRiP_median <- obj_filtered[["FRiP"]] %>% pull(FRiP) %>% median()
cellFRiP_mean <- obj_filtered[["FRiP"]] %>% pull(FRiP) %>% mean()
stats <- stats %>% 
    add_row(Term = "cellFrag_median", value = as.character(cellFrag_median), .after = 2) %>%
    add_row(Term = "cellFrag_mean", value = as.character(cellFrag_mean), .after = 3) %>%
    add_row(Term = "cellFRiP_median", value = as.character(cellFRiP_median), .after = 4) %>%
    add_row(Term = "cellFRiP_mean", value = as.character(cellFRiP_mean), .after=5)
```

Analysis Result {data-orientation=rows}
====================


<h2>Sample: `r params$sampleName`</h2>

Row
----------

### Estimated Number of Cells

```{r}
valueBox(scales::label_comma()(length(cells)), caption = "Estimated number of cells", icon="fa-circle-notch", color="primary")
```

### Median of Total Fragments in Cells

```{r}
valueBox(scales::label_comma()(cellFrag_median), caption = "Median fragments per cell", icon="fa-chart-bar", color="danger")
```

### Median FRiP of Cells

```{r}
valueBox(scales::label_percent(0.01)(cellFRiP_median), caption = "Median fraction of fragments overlapping peaks", icon="fa-barcode", color="info")
```

Row
----------

### Cells

```{r}

tribble(
  ~"", ~"",
  "Estimated number of cells", scales::label_comma()(length(cells)),
  "Mean fragments per cell", scales::label_comma()(cellFrag_mean),
  "Median fragments per cell", scales::label_comma()(cellFrag_median),
  "Mean of FRiP in cells", scales::label_percent(0.01)(cellFRiP_mean),
  "Median of FRiP in cells", scales::label_percent(0.01)(cellFRiP_median)
  ##"Fraction of transposition events in peaks in cells", "100",
) %>% kbl() %>% kable_paper("hover", html_font = "arial, helvetica, sans-serif") %>%
  column_spec(1, bold = T)
```

### Sequencing

```{r}
valid_barcode_reads <- stats %>% filter(Term=="valid_barcode_reads") %>% pull(value) %>% as.numeric()
total_input_reads <- stats %>% filter(Term=="total_input_reads") %>% pull(value) %>% as.numeric()
me_containing_reads <- stats %>% filter(Term == "me_containing_reads") %>% pull(value) %>% as.numeric()
me_read_ratio <- me_containing_reads/total_input_reads
valid_barcode_ratio <- valid_barcode_reads/total_input_reads
barcode_q30_ratio <- stats %>% filter(Term == "barcode_q30_ratio") %>% pull(value) %>% as.numeric()
r1_q30_ratio <- stats %>% filter(Term=="r1_q30_ratio") %>% pull(value) %>% as.numeric()
r2_q30_ratio <- stats %>% filter(Term=="r2_q30_ratio") %>% pull(value) %>% as.numeric()

tribble(
  ~"", ~"",
  "Raw read pairs", label_comma()(total_input_reads),
  "Valid barcodes", label_percent(0.01)(valid_barcode_ratio),
  "ME containing read pairs", label_percent(0.01)(me_read_ratio),
  "Q30 bases in barcode", label_percent(0.01)(barcode_q30_ratio),
  "Q30 bases in read 1", label_percent(0.01)(r1_q30_ratio),
  "Q30 bases in read 2", label_percent(0.01)(r2_q30_ratio)
) %>% kbl() %>% kable_paper("hover", html_font = "arial, helvetica, sans-serif") %>%
  column_spec(1, bold = T)
```

Row {data-height=500}
----------

### Cells Knee Plot

```{r}
fig <- plot_ly(data = d, x = ~idx, y = ~fragsOverlapPeaks, type = 'scatter', mode = 'lines',
               color = ~group, colors = c('grey', '#002366'),
               hovertemplate = 'Fragments: %{y}<extra>%{fullData.name}</extra>',
               line = list(width = 3),
               height = 400, width = 360)
fig <- layout(fig, xaxis = list(type = "log", title = "Barcodes"), 
              yaxis = list(type = "log", title = "Fragments in Peaks"), showlegend = FALSE)
config_plotly_fig(fig) %>% toWebGL()
```

### Cells Fragments Distribution

```{r}
p <- ggplot(d, aes(fragments, color=group, fill=group)) + geom_density(alpha=0.4, adjust = 5)+
  scale_x_log10() + xlab("Fragments") + theme_bw() + 
  theme(panel.grid=element_blank(), legend.position=c(0.8,0.9), legend.title=element_blank()) +
  scale_fill_manual(values=c("grey", "steelblue")) +
  scale_color_manual(values=c("grey", "steelblue"))
ggplotly(p) %>% config_plotly_fig() %>% toWebGL()
```

Row {data-height=600}
----------

### Cells Metrics

```{r fig.width=15, fig.height=7, fig.fullwidth=TRUE, out.width="100%"}
qcPlot
```

Row {data-height=600}
----------

### Cluster (Colored by Group)

```{r}
if(length(cells)>=150){
  clusterPlot %>% ggplotly() %>% config_plotly_fig() %>% toWebGL()
}else{
  cat("Too few cells\n")
}
```

### Cluster (Colored by Depth)

```{r}
if(length(cells)>=150){
  obj_filtered[["logFragments"]] <- log10(obj_filtered[["fragments"]])
  depthPlot <- FeaturePlot(object = obj_filtered, features = 'logFragments') + scale_color_continuous(name="log10 Fragments", type = "viridis", direction = 1) + labs(title = NULL)
  depthPlot %>% ggplotly() %>% config_plotly_fig() %>% toWebGL()
}else{
  cat("Too few cells\n")
}
```

Data Quality {data-orientation=rows}
====================

Row {data-height=500}
----------

### Library Complexity

```{r echo=FALSE}
pbc1 <- stats %>% filter(Term=="pbc1") %>% pull(value) %>% as.numeric()
pbc2 <- stats %>% filter(Term=="pbc2") %>% pull(value) %>% as.numeric()
nrf <- stats %>% filter(Term=="nrf") %>% pull(value) %>% as.numeric()
duplicationReads <- stats %>% filter(Term=="duplicated_reads") %>% pull(value) %>% as.numeric()
duplicationRatio <- duplicationReads/me_containing_reads
tribble(
  ~"", ~"",
  "PBC1", label_number(0.01)(pbc1),
  "PBC2", label_number(0.01)(pbc2),
  "NRF", label_number(0.01)(nrf),
  "Duplication", label_percent(0.01)(duplicationRatio)
) %>%
kbl() %>% 
kable_paper("hover", html_font = "arial, helvetica, sans-serif") %>%
column_spec(1, bold = T)
```

### Sequencing Saturation

```{r}
if(!is.null(params$saturation_json)){
  saturation_data <- fromJSON(params$saturation_json) %>%
      as_tibble() %>%
      mutate(across(everything(), .fns = as.numeric)) %>%
      mutate(saturation = 1-uniqueFrags/seqDep) %>%
      mutate(saturation = scales::label_percent(0.01)(saturation))
}

plot_ly(saturation_data, x= ~seqDep, y= ~uniqueFrags, text= ~saturation,
       type = 'scatter', mode = 'lines',
       hovertemplate = 'Read Pairs: %{x}<br>Unique Fragments: %{y}<br>Saturation: %{text}<extra></extra>') %>%
    layout(xaxis = list(title = 'Sequencing Read Depth'), yaxis = list(title = 'Total Fragments')) %>%
    config_plotly_fig() %>%
    toWebGL()
```

Row
----------

### Mapping Result

```{r}
total_mapped_reads <- stats %>% filter(Term == "total_mapped_reads") %>% pull(value) %>% as.numeric()
total_mapping_rate <- total_mapped_reads/me_containing_reads
q30_mapped_reads <- stats %>% filter(Term == "q30_mapped_reads") %>% pull(value) %>% as.numeric()
q30_mapping_rate <- q30_mapped_reads/me_containing_reads
total_frags <- stats %>% filter(Term == "total_frags") %>% pull(value) %>% as.numeric()
mito_frags <- stats %>% filter(Term == "mito_frags") %>% pull(value) %>% as.numeric()
nfr_frags <- stats %>% filter(Term == "nfr_frags") %>% pull(value) %>% as.numeric()
mono_frags <- stats %>% filter(Term == "mono_frags") %>% pull(value) %>% as.numeric()

tribble(
  ~"", ~"",
  "Total mapped read pairs", scales::label_comma()(total_mapped_reads),
  "Total mapping rate", label_percent(0.01)(total_mapping_rate),
  "High confident mapping rate", label_percent(0.01)(q30_mapping_rate),
  "Total fragments (after dedup)", scales::label_comma()(total_frags),
  "Mitochondrial fragments", scales::label_comma()(mito_frags),
  "Fragments in nucleosome-free regions", scales::label_comma()(nfr_frags),
  "Fragments in mono-nucleosome regions", scales::label_comma()(mono_frags),
) %>% kbl() %>% kable_paper("hover", html_font = "arial, helvetica, sans-serif") %>%
  column_spec(1, bold = T)
```

### Insert Size

```{r}
fragLenPlot %>% ggplotly() %>% config_plotly_fig() %>% toWebGL()
```

Row
----------

### Peaks

```{r}
tss_score <- obj_filtered@assays$peaks@positionEnrichment$TSS %>% apply(2, mean) %>% max()
raw_peaks_num <- read_tsv(params$macs_peaks) %>% nrow
tribble(
  ~"", ~"",
  "Raw macs peaks", scales::label_comma()(raw_peaks_num),
  "Peaks observed in cells", scales::label_comma()(nrow(obj_filtered)),
  "TSS enrichment score", scales::label_number(0.01)(tss_score)
) %>% kbl() %>% kable_paper("hover", html_font = "arial, helvetica, sans-serif") %>%
  column_spec(1, bold = T)
```

### Peaks Targeting

```{r}
## Some points have FRiP larger than 1, needs to be checked
ggplot(d, aes(x=fragments, y=FRiP, color=group))+geom_point(alpha=0.8)+scale_x_log10() + 
    theme_bw() + scale_color_manual(values=c("grey", "steelblue")) + ylim(0,1) +
    theme(legend.title=element_blank(), panel.grid=element_blank())
##ggplotly(p) %>% toWebGL()
```

Row
----------

### TSS Enrichment Signal

```{r}
tssPlot %>% ggplotly() %>% config_plotly_fig() %>% toWebGL()
```

### TSS Enrichment Distribution

```{r}
## Some points have FRiP larger than 1, needs to be checked
ggplot(d, aes(x=fragments, y=TSS.enrichment, color=group))+geom_point(alpha=0.8)+scale_x_log10() + 
    theme_bw() + scale_color_manual(values=c("grey", "steelblue")) + ylim(0,1) +
    theme(legend.title=element_blank(), panel.grid=element_blank())
```

```{js}
// set knee plot margin to auto
$( document ).ready(function() {
    document.querySelector(".plot-container").querySelector(".svg-container").style.margin = "auto";
});
```

```{r include=FALSE}
stats <- stats %>% 
    add_row(Term = "total_mapping_rate", value = as.character(total_mapping_rate), .after = 10) %>%
    add_row(Term = "q30_mapping_rate", value = as.character(q30_mapping_rate), .after = 13) %>%
    add_row(Term = "valid_barcode_ratio", value = as.character(valid_barcode_ratio), .after = 8)

write_tsv(stats, file = paste0(params$sampleName, "_combined_stats.tsv"))
```
