---
title: scATAC-seq Report
date: "`r Sys.Date()`"
output:
  flexdashboard::flex_dashboard:
    vertical_layout: scroll
    orientation: rows
    theme:
      version: 5
      bootswatch: sandstone
      bslib: true
      base_font:
        google: Asap
params:
  nCPUs: 6
  nMem: 10000000000.0
  sampleName: NULL
  stats_tsv: NULL
  raw_cells: NULL
  raw_meta: NULL
  obj_filtered: NULL
  macs_peaks: NULL
  saturation_json: NULL
  version_json: NULL
---

```{r include=FALSE}
library(tidyverse)
library(scales)
library(kableExtra)
library(flexdashboard)
library(Signac)
library(Seurat)
##library(GenomeInfoDb)
##library(EnsDb.Mmusculus.v79)
library(rtracklayer)
library(patchwork)
library(future)
library(plotly)
library(htmltools)
library(jsonlite)
library(reticulate)

set.seed(1234)
plan("multisession", workers = as.numeric(params$nCPUs))
options(future.globals.maxSize = as.numeric(params$nMem))
RhpcBLASctl::blas_set_num_threads(1) # https://github.com/satijalab/seurat/issues/3991


## Rmd options
knitr::opts_chunk$set(
  comment = '', fig.retina = 4,
  warning = FALSE, message = FALSE
)

config_plotly_fig <- function(fig){
    config(
        fig,
        displaylogo = FALSE, 
        modeBarButtonsToRemove = c('zoom', 'pan', 'select', 'zoomIn', 'zoomOut', 'autoScale',
                                   'hoverClosestCartesian', 'hoverCompareCartesian'),
        toImageButtonOptions = list(height= NULL, width= NULL, scale= 2)
    )
}
```

```{bslib}
.chart-title {
  font-size: 1.2rem;
  font-weight: 700;
}
```

```{js echo=FALSE}
document.querySelector(".navbar").classList.remove("navbar-inverse");
document.querySelector(".navbar").classList.add("navbar-default");
document.querySelector(".navbar").classList.add("pt-1");
document.querySelector(".navbar").classList.add("pb-1");
```

```{r infoIcon, include=FALSE}
infoIcon <- function(targetEl, size = "1.2rem"){
  querySelector <- paste0("#", targetEl)
  ## Make HTML
  questionIconString <- paste0(
      '<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16" class="bi bi-question-circle-fill " style="height:',
      size,
      ';width:',
      size,
      ';fill:currentColor;vertical-align:-0.125em;fill:var(--bs-primary);" aria-hidden="true" role="img"><path d="M16 8A8 8 0 1 1 0 8a8 8 0 0 1 16 0zM5.496 6.033h.825c.138 0 .248-.113.266-.25.09-.656.54-1.134 1.342-1.134.686 0 1.314.343 1.314 1.168 0 .635-.374.927-.965 1.371-.673.489-1.206 1.06-1.168 1.987l.003.217a.25.25 0 0 0 .25.246h.811a.25.25 0 0 0 .25-.25v-.105c0-.718.273-.927 1.01-1.486.609-.463 1.244-.977 1.244-2.056 0-1.511-1.276-2.241-2.673-2.241-1.267 0-2.655.59-2.75 2.286a.237.237 0 0 0 .241.247zm2.325 6.443c.61 0 1.029-.394 1.029-.927 0-.552-.42-.94-1.029-.94-.584 0-1.009.388-1.009.94 0 .533.425.927 1.01.927z"></path></svg>'
  )
  outTag <- tags$i(
      HTML(questionIconString),
      `data-bs-toggle`="modal",
      `data-bs-target`=querySelector
  )
  return(outTag)
}

createModal <- function(Id, expl_tbl){
    contentTag <- expl_tbl %>%
        apply(1, FUN=function(x){
            tagList(
                h6(HTML(x[1]), style="font-weight:700;"),
                p(HTML(x[2]))
            )
        }) %>% 
        tagList()

   outTag <- div(
       class = "modal fade",
       id = Id,
       tabindex = -1,
       `aria-labelledby`="exampleModalLabel",
       `aria-hidden`="true",
       div(
           class = "modal-dialog modal-dialog-centered modal-dialog-scrollable modal-lg",
           div(
               class = "modal-content",
               div(
                   class = "modal-body",
                   div(
                       class = "explanation",
                       contentTag
                   )
               )
           )
       )
   )
   return(outTag)
}
```

```{css}
.explanation {
  background-color: rgba(var(--bs-secondary-rgb), 0.2); 
  border-radius: var(--bs-modal-border-radius);
  padding: 1rem
}
```

```{r modal, echo = FALSE}
expl_tbl <- tribble(
    ~title, ~content,
    "Estimated number of cells",
    "Number of barcodes identified as cells by otsu's method",
    "Mean fragments per cell",
    "The mean value of high-quality fragments per cell. The high-quality fragments were generated by deduplicating valid barcode reads which were mapped to genome with mapping quality > 30",
    "Median fragments per cell",
    "The median value of the high-quality fragments per cell",
    "Mean of FRiP (percentage of fragments in peaks) in cells",
    "The mean value of the high-quality fragments overlapping peaks in cells",
    "Median of FRiP (percentage of fragments in peaks) in cells",
    "The median value of the high-quality fragments overlapping peaks in cells"
)
createModal("cellStats", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Raw read pairs",
    "Total number of raw read pairs sequenced",
    "Valid barcodes",
    "The fraction of reads with valid barcode. The raw barcode sequence were matched to whitelist allowing 1 nt mismatch for 10X scATAC data. For ThunderBio scATAC data, 1 nt mismatch is allowed for each barcode segments.",
    "ME containing read pairs",
    "The fraction of the valid barcode read pairs with ME sequence also detected",
    "Q30 bases in barcode",
    "Percentage of the nucleotides in cell barcode with the sequencing quality larger than 30 (99.9% accuracy)",
    "Q30 bases in read 1",
    "Percentage of the nucleotides in read 1 with sequencing quality larger than 30 (99.9% accuracy)",
    "Q30 bases in read 2",
    "Percentage of the nucleotides in read 2 with sequencing quality larger than 30 (99.9% accuracy)"
)
createModal("seqStats", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Cells Knee Plot",
    'The Knee Plot is also called Barcode Rank Plot. The graph displays the number of fragments overlapping peaks assigned to each barcode. The blue color indicates cell associated barcodes, while grey ones are background barcodes. Cell associated barcodes were determined by <a href="https://en.wikipedia.org/wiki/Otsu%27s_method">otsu\'s method</a> followed by <a href="https://doi.org/10.1186/s13059-019-1662-y">EmptyDrops test</a>.'
)
createModal("kneePlot", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Cells Fragments Distribution",
    "This plot illustrates the fragment counts distribution in cell associated barcodes and background barcodes."
)
createModal("fragmentsDistribution", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Cells Metrics",
    "The violin plots summarizing cells metrics including: <li>FRiP: percentage of fragments in peaks</li><li>nCount_peaks: counts of the fragments in peaks</li><li>TSS.enrichment: TSS signal enrichment</li><li>blacklist_fraction: counts of the fragments located in blacklist regions</li><li>nucleosome_signal: ratio of fragments between 147 bp and 294 bp (mononucleosome) to fragments < 147 bp (nucleosome-free))</li>"
)
createModal("cellMetrics", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Cell Cluster (Colored By Group)",
    'The UMAP dimension reduction result of the cellular open chromatin region profile. Cell-peaks matrix was processed by latent semantic indexing (LSI) method implemented in <a href="https://stuartlab.org/signac/articles/pbmc_vignette#normalization-and-linear-dimensional-reduction">Signac</a> package, and clustered by  graph-based clustering method implemented in Seurat package. Cells were colored by cell clusters. Please note we also discarded features observed in less than 10 cells and cells with less than 200 features.'
)
createModal("clusterCellGroups", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Cell Cluster (Colored By Depth)",
    'The UMAP dimension reduction result of the cellular open chromatin region profile. Cell-peaks matrix was processed by latent semantic indexing (LSI) method implemented in <a href="https://stuartlab.org/signac/articles/pbmc_vignette#normalization-and-linear-dimensional-reduction">Signac</a> package, and clustered by  graph-based clustering method implemented in Seurat package. Cells were colored by fragment counts (log-transformed). Please note we also discarded features observed in less than 10 cells and cells with less than 200 features.'
)
createModal("clusterFragmentCounts", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "PBC1",
    "PCR Bottlenecking Coefficient 1 (PBC1): PBC1=M<sub>1</sub>/M<sub>DISTINCT</sub> where M<sub>1</sub> is the number of fragments corresponding to only one read pair; M<sub>DISTINCT</sub> is the number of total fragments after deduplication",
    "PBC2",
    "PCR Bottlenecking Coefficient 2 (PBC2): PBC2=M<sub>1</sub>/M<sub>2</sub> where M<sub>1</sub> is the number of fragments corresponding to only one read pair, M<sub>2</sub> is the number of fragments corresponding to two read pairs",
    "NRF",
    "Non-Redundant Fraction (NRF): Number of fragments (i.e. after removing duplicates) / Total number of reads.",
    "Note",
    'These metrics were widely used in bulk ATAC experiments to measure the library complexity, please ref to <a href="https://www.encodeproject.org/data-standards/terms/">ENCODE website</a> for detail explanation and suggested standards.'
)
createModal("libraryComplexity", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Sequencing Saturation",
    "The reads of the library were downsampled at different ratio to calculate the sequencing saturation at different sequencing depth. The saturation was calculated as 1-nFragment/nReadPairs, where nFragment is the fragments in cells."
)
createModal("saturationCurve", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Total mapped read pairs",
    "Total number of the read pairs mapped to the genome with valid barcode and ME sequence.",
    "Total mapping rate",
    "Fraction of the read pairs mapped to the genome: total_mapped_read_pairs/me_containing_read_pairs.",
    "High confident mapping rate",
    "Fraction of the read pairs with mapping quality > 30, which were used to generate fragments subsequenctly.",
    "Total fragments (after dedup)",
    "Number of high-quality fragments generated by deduplicating high-confident mapped read pairs.",
    "Mitochondrial fragments",
    "Number of fragments from mitochondrial genome.",
    "Fragments in nucleosome-free regions",
    "Number of fragments from nucleosome-free regions (length < 147 bp).",
    "Fragments in mono-nucleosome regions",
    "Number of fragments from mono-nucleosome regions (length > 147 bp, < 147*2 bp)."
)
createModal("mappingStats", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Insert Size",
    "Histogram showing fragments size distribution. NS is nucleosome signal, which computes the ratio of mononucleosome fragments to nucleosome-free fragments."
)
createModal("fragmentsSize", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Raw macs2 peaks",
    "Total number of peaks generated by pseudobulk peak calling using macs2.",
    "Peaks observed in cells",
    "Number of peaks observed in the cell associated barcodes.",
    "TSS enrichment score",
    'Maximum value of the transcription start site (TSS) signal enrichment. The enrichment score was calculated by summarizing normalized read depth around reference TSS, please refer to <a href="https://www.encodeproject.org/data-standards/terms/">ENCODE website</a> for details. Here we used 2000 bp window (upstream 1000 bp + downstream 1000 bp) to keep the score consistent with cellranger report. Here we grouped cells according to TSS enrichment score, the "high" group profiled all TSS sites of the cells with enrichment score larger than 3.'
)
createModal("peaks", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Peaks Targeting",
    "Scatter plot represents FRiP distribution among all barcodes, blue ones are cell associated barcodes while grey ones indicate backaground barcodes."
)
createModal("fripScatter", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "TSS Enrichment Signal",
    'Typical TSS enrichment plot, summarizing normalized read depth around reference TSS, please refer to <a href="https://www.encodeproject.org/data-standards/terms/">ENCODE website</a> for details.'
)
createModal("tssPlot", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Peaks Targeting",
    "Scatter plot represents TSS enrichment score distribution among all barcodes, blue ones are cell associated barcodes while grey ones indicate backaground barcodes."
)
createModal("tssScatter", expl_tbl)
```

```{r include=FALSE}
html_font <- "Asap, Arial, sans-serif"
```

```{r preprocess, include=FALSE}
obj_filtered <- readRDS(params$obj_filtered)
obj_filtered$orig.ident <- params$sampleName

tssPlot <- TSSPlot(obj_filtered, group.by = 'high.tss') + NoLegend()

## Chr1 fragments length distribution
fragLenPlot <- FragmentHistogram(object = obj_filtered, group.by = 'nucleosome_group', region=paste0(seqlevels(Annotation(obj_filtered))[1],'-1-200000000'))

qcPlot <- VlnPlot(
  object = obj_filtered,
  features = c('FRiP', 'nCount_peaks',
               'TSS.enrichment', 'blacklist_fraction', 'nucleosome_signal'),
  pt.size = 0.1,
  ncol = 5,
  group.by = "orig.ident"
)


## Filter data with otsu threshold of nCount_peaks, and with other naive criterias
cells <- colnames(obj_filtered)

raw_cells <- read_tsv(params$raw_cells, col_names = FALSE) %>% pull()

d <- read_tsv(params$raw_meta) ## counts = fragsOverlapPeaks

d <- d %>% dplyr::arrange(desc(fragsOverlapPeaks)) %>%
    dplyr::mutate(idx=1:nrow(d), group = factor(group, levels = c("cell", "background")))

## Do not perform clustering when there are too few cells
if(length(cells)>=150){
    clusterPlot <- DimPlot(object = obj_filtered, label = TRUE)
}
```

```{r read_stats, include = FALSE}
stats <- read_tsv(params$stats_tsv, col_names=FALSE)
colnames(stats) <- c("Term", "value")
## Add more stats about cells
cellFrag_median <- obj_filtered[["fragments"]] %>% pull(fragments) %>% median() %>% as.integer()
cellFrag_mean <- obj_filtered[["fragments"]] %>% pull(fragments) %>% mean() %>% as.integer()
cellFRiP_median <- obj_filtered[["FRiP"]] %>% pull(FRiP) %>% median()
cellFRiP_mean <- obj_filtered[["FRiP"]] %>% pull(FRiP) %>% mean()
stats <- stats %>% 
    add_row(Term = "cellFrag_median", value = as.character(cellFrag_median), .after = 2) %>%
    add_row(Term = "cellFrag_mean", value = as.character(cellFrag_mean), .after = 3) %>%
    add_row(Term = "cellFRiP_median", value = as.character(cellFRiP_median), .after = 4) %>%
    add_row(Term = "cellFRiP_mean", value = as.character(cellFRiP_mean), .after=5)
```

Analysis Result {data-orientation=rows}
====================


<h2>Sample: `r params$sampleName`</h2>

Row
----------

### Estimated Number of Cells

```{r}
valueBox(scales::label_comma()(length(raw_cells)), caption = "Estimated number of cells", icon="fa-circle-notch", color="primary")
```

### Median of Total Fragments in Cells

```{r}
valueBox(scales::label_comma()(cellFrag_median), caption = "Median fragments per cell", icon="fa-chart-bar", color="danger")
```

### Median FRiP of Cells

```{r}
valueBox(scales::label_percent(0.01)(cellFRiP_median), caption = "Median fraction of fragments overlapping peaks", icon="fa-barcode", color="info")
```

Row
----------

### Cell Statistics `r infoIcon("cellStats")`

```{r}

statsOut1 <- tribble(
  ~"", ~"",
  "Estimated number of cells", scales::label_comma()(length(raw_cells)),
  "Mean fragments per cell", scales::label_comma()(cellFrag_mean),
  "Median fragments per cell", scales::label_comma()(cellFrag_median),
  "Mean of FRiP in cells", scales::label_percent(0.01)(cellFRiP_mean),
  "Median of FRiP in cells", scales::label_percent(0.01)(cellFRiP_median)
  ##"Fraction of transposition events in peaks in cells", "100",
)

statsOut1 %>%
  kbl() %>%
  kable_paper("hover", html_font = html_font, font_size = "1rem") %>%
  column_spec(1, bold = T)
```

### Sequencing Statistics `r infoIcon("seqStats")`

```{r}
valid_barcode_reads <- stats %>% filter(Term=="valid_barcode_reads") %>% pull(value) %>% as.numeric()
total_input_reads <- stats %>% filter(Term=="total_input_reads") %>% pull(value) %>% as.numeric()
me_containing_reads <- stats %>% filter(Term == "me_containing_reads") %>% pull(value) %>% as.numeric()
me_read_ratio <- me_containing_reads/total_input_reads
valid_barcode_ratio <- valid_barcode_reads/total_input_reads
barcode_q30_ratio <- stats %>% filter(Term == "barcode_q30_ratio") %>% pull(value) %>% as.numeric()
r1_q30_ratio <- stats %>% filter(Term=="r1_q30_ratio") %>% pull(value) %>% as.numeric()
r2_q30_ratio <- stats %>% filter(Term=="r2_q30_ratio") %>% pull(value) %>% as.numeric()

statsOut2 <- tribble(
  ~"", ~"",
  "Raw read pairs", label_comma()(total_input_reads),
  "Valid barcodes", label_percent(0.01)(valid_barcode_ratio),
  "ME containing read pairs", label_percent(0.01)(me_read_ratio),
  "Q30 bases in barcode", label_percent(0.01)(barcode_q30_ratio),
  "Q30 bases in read 1", label_percent(0.01)(r1_q30_ratio),
  "Q30 bases in read 2", label_percent(0.01)(r2_q30_ratio)
)

statsOut2 %>%
  kbl() %>%
  kable_paper("hover", html_font = html_font, font_size = "1rem") %>%
  column_spec(1, bold = T)
```

Row {data-height=500}
----------

### Cells Knee Plot `r infoIcon("kneePlot")`

```{r}
d0 <- d %>% mutate(group = factor(group, levels = c("background", "cell")))
fig <- plot_ly(data = d0, x = ~idx, y = ~fragsOverlapPeaks, type = 'scatter', mode = 'lines',
               color = ~group, colors = c('grey', '#002366'),
               hovertemplate = 'Fragments: %{y}<extra>%{fullData.name}</extra>',
               line = list(width = 3),
               height = 400, width = 360)
fig <- layout(fig, xaxis = list(type = "log", title = "Barcodes"), 
              yaxis = list(type = "log", title = "Fragments in Peaks"), showlegend = FALSE)
config_plotly_fig(fig) %>% toWebGL()
```

### Cells Fragments Distribution `r infoIcon("fragmentsDistribution")`

```{r}
p <- ggplot(d, aes(fragments, color=group, fill=group)) + geom_density(alpha=0.4, adjust = 5)+
  scale_x_log10(labels = label_number(scale_cut = cut_short_scale())) + 
  xlab("Fragments") + 
  theme_bw() + 
  theme(panel.grid=element_blank(), legend.position=c(0.8,0.9), legend.title=element_blank()) +
  scale_fill_manual(values=c("steelblue", "grey")) +
  scale_color_manual(values=c("steelblue", "grey"))
ggplotly(p) %>% config_plotly_fig() %>% toWebGL()
```

Row {data-height=600}
----------

### Cells Metrics `r infoIcon("cellMetrics")`

```{r fig.width=15, fig.height=7, fig.fullwidth=TRUE, out.width="100%"}
qcPlot
```

Row {data-height=600}
----------

### Cell Cluster (Colored By Group) `r infoIcon("clusterCellGroups")`

```{r}
if(length(cells)>=150){
  clusterPlot %>% ggplotly() %>% config_plotly_fig() %>% toWebGL()
}else{
  cat("Too few cells\n")
}
```

### Cell Cluster (Colored By Depth) `r infoIcon("clusterFragmentCounts")`

```{r}
if(length(cells)>=150){
  obj_filtered[["logFragments"]] <- log10(obj_filtered[["fragments"]])
  depthPlot <- FeaturePlot(object = obj_filtered, features = 'logFragments') + scale_color_continuous(name="log10 Fragments", type = "viridis", direction = 1) + labs(title = NULL)
  depthPlot %>% ggplotly() %>% config_plotly_fig() %>% toWebGL()
}else{
  cat("Too few cells\n")
}
```

Data Quality {data-orientation=rows}
====================

Row {data-height=500}
----------

### Library Complexity `r infoIcon("libraryComplexity")`

```{r echo=FALSE}
pbc1 <- stats %>% filter(Term=="pbc1") %>% pull(value) %>% as.numeric()
pbc2 <- stats %>% filter(Term=="pbc2") %>% pull(value) %>% as.numeric()
nrf <- stats %>% filter(Term=="nrf") %>% pull(value) %>% as.numeric()
duplicationReads <- stats %>% filter(Term=="duplicated_reads") %>% pull(value) %>% as.numeric()
duplicationRatio <- duplicationReads/me_containing_reads

statsOut3 <- tribble(
  ~"", ~"",
  "PBC1", label_number(0.01)(pbc1),
  "PBC2", label_number(0.01)(pbc2),
  "NRF", label_number(0.01)(nrf),
  "Duplication", label_percent(0.01)(duplicationRatio)
)

statsOut3 %>%
  kbl() %>% 
  kable_paper("hover", html_font = html_font, font_size = "1rem") %>%
  column_spec(1, bold = T)
```

### Sequencing Saturation `r infoIcon("saturationCurve")`

```{r}
if(!is.null(params$saturation_json)){
  saturation_data <- fromJSON(params$saturation_json) %>%
      as_tibble() %>%
      mutate(across(everything(), .fns = as.numeric)) %>%
      mutate(saturation = 1-total_frags/total_reads) %>%
      mutate(saturation = scales::label_percent(0.01)(saturation))
}

plot_ly(saturation_data, x= ~total_reads, y= ~medianFragsInCells, text= ~saturation,
       type = 'scatter', mode = 'lines',
       hovertemplate = 'Read Pairs: %{x}<br>Median of Fragments in Peaks: %{y}<br>Saturation: %{text}<extra></extra>') %>%
    layout(xaxis = list(title = 'Sequencing Read Depth'), yaxis = list(title = 'Median Fragments Overlapping Peaks in Cells')) %>%
    config_plotly_fig() %>%
    toWebGL()
```

Row
----------

### Mapping Result `r infoIcon("mappingStats")`

```{r}
total_mapped_reads <- stats %>% filter(Term == "total_mapped_reads") %>% pull(value) %>% as.numeric()
total_mapping_rate <- total_mapped_reads/me_containing_reads
q30_mapped_reads <- stats %>% filter(Term == "q30_mapped_reads") %>% pull(value) %>% as.numeric()
q30_mapping_rate <- q30_mapped_reads/me_containing_reads
total_frags <- stats %>% filter(Term == "total_frags") %>% pull(value) %>% as.numeric()
mito_frags <- stats %>% filter(Term == "mito_frags") %>% pull(value) %>% as.numeric()
nfr_frags <- stats %>% filter(Term == "nfr_frags") %>% pull(value) %>% as.numeric()
mono_frags <- stats %>% filter(Term == "mono_frags") %>% pull(value) %>% as.numeric()

statsOut4 <- tribble(
  ~"", ~"",
  "Total mapped read pairs", scales::label_comma()(total_mapped_reads),
  "Total mapping rate", label_percent(0.01)(total_mapping_rate),
  "High confident mapping rate", label_percent(0.01)(q30_mapping_rate),
  "Total fragments (after dedup)", scales::label_comma()(total_frags),
  "Mitochondrial fragments", paste0(scales::label_comma()(mito_frags), " (", label_percent(0.01)(mito_frags/total_frags),")"),
  "Fragments in nucleosome-free regions", paste0(scales::label_comma()(nfr_frags), " (", label_percent(0.01)(nfr_frags/total_frags),")"),
  "Fragments in mono-nucleosome regions", paste0(scales::label_comma()(mono_frags), " (", label_percent(0.01)(mono_frags/total_frags),")")
)

statsOut4 %>% kbl() %>% kable_paper("hover", html_font = html_font, font_size = "1rem") %>%
  column_spec(1, bold = T)
```

### Insert Size `r infoIcon("fragmentsSize")`

```{r}
fragLenPlot %>% ggplotly() %>% config_plotly_fig() %>% toWebGL()
```

Row
----------

### Peaks `r infoIcon("peaks")`

```{r}
tss_score <- obj_filtered@assays$peaks@positionEnrichment$TSS %>% apply(2, mean) %>% max()
raw_peaks_num <- read_tsv(params$macs_peaks) %>% nrow
statsOut5 <- tribble(
  ~"", ~"",
  "Raw macs2 peaks", scales::label_comma()(raw_peaks_num),
  "Peaks observed in cells", scales::label_comma()(nrow(obj_filtered)),
  "TSS enrichment score", scales::label_number(0.01)(tss_score)
)

statsOut5 %>%
  kbl() %>%
  kable_paper("hover", html_font = html_font, font_size = "1rem") %>%
  column_spec(1, bold = T)
```

### Peaks Targeting `r infoIcon("fripScatter")`

```{r}
## Some points have FRiP larger than 1, needs to be removed
ggplot(d, aes(x=fragments, y=FRiP, color=group))+geom_point(alpha=0.5, size=0.4)+scale_x_log10() + 
    theme_bw() + scale_color_manual(values=c("steelblue", "grey")) + ylim(0,1) +
    theme(legend.title=element_blank(), panel.grid=element_blank())
##ggplotly(p) %>% toWebGL()
```

Row
----------

### TSS Enrichment Signal `r infoIcon("tssPlot")`

```{r}
tssPlot %>% ggplotly() %>% config_plotly_fig() %>% toWebGL()
```

### TSS Enrichment Distribution `r infoIcon("tssScatter")`

```{r}
ggplot(d, aes(x=fragments, y=TSS.enrichment, color=group))+geom_point(alpha=0.5, size=0.4)+scale_x_log10() + 
    theme_bw() + scale_color_manual(values=c("steelblue", "grey")) +
    theme(legend.title=element_blank(), panel.grid=element_blank())
```

Running Info {data-orientation=rows}
====================

Row
----------

### Running Information

```{r}
version_info <- fromJSON(params$version_json)
version_tbl <- tribble(
  ~term, ~value,
  "Sample ID", params$sampleName,
  "Pipeline Version", version_info$pipeline_version,
  "Reference Genome", version_info$referenceGenome,
  "Reference GTF", version_info$referenceGTF,
  "BWA Index", version_info$bwa_index,
  "BWA Version", version_info$bwa_version
)
names(version_tbl) <- c("", "")
version_tbl %>% 
    kbl() %>% 
    kable_paper("hover", html_font = html_font)
```

```{js}
$( document ).ready(function() {
    // set knee plot margin to auto
    document.querySelector(".plot-container .svg-container").style.margin = "auto";
});
```

```{r include=FALSE}
stats <- stats %>% 
    add_row(Term = "total_mapping_rate", value = as.character(total_mapping_rate), .after = 10) %>%
    add_row(Term = "q30_mapping_rate", value = as.character(q30_mapping_rate), .after = 13) %>%
    add_row(Term = "valid_barcode_ratio", value = as.character(valid_barcode_ratio), .after = 8)

write_tsv(stats, file = paste0(params$sampleName, "_combined_stats.tsv.gz"))
```

```{r include=FALSE}
statsOut <- bind_rows(
  statsOut1,
  statsOut2,
  statsOut3,
  statsOut4,
  statsOut5
)
colnames(statsOut) <- c("item", "value")
statsOut %>%
  add_row(item = "sampleID", value = params$sampleName, .before = 1) %>%
  pivot_wider(names_from=item, values_from=value) %>%
  write_json(paste0(params$sampleName, ".metrics.json"), pretty=TRUE)
```
